### PR TITLE
<format>: all nans are "nan"

### PR DESCRIPTION
**Pull request by CaseyCarter**
... per [[format.string.std]/22](http://eel.is/c++draft/format#string.std-22).

Reported by Mark de Wever after we tried to break their perfectly good test with https://reviews.llvm.org/D131336.

### CODE REVIEW

File: stl/inc/format [hunk 1]
```diff
@@ -2773,19 +2790,30 @@ _NODISCARD _OutputIt _Fmt_write(
...
+
+    if ((_STD isnan)(_Value)) {
+        _Result.ptr = _Buffer;
+        if (_Is_negative) {
+            ++_Result.ptr; // pretend to skip over a '-' that to_chars would put in _Buffer[0]
+        }
+
+        memcpy(_Result.ptr, "nan", 3);
+        _Result.ptr += 3;
+        _Result.ec = errc{};
```
Comment (CaseyCarter) [inline stl/inc/format:2803]: This is unnecessary since you moved the `ec` check into the `else` branch.

File: stl/inc/format [hunk 2]
```diff
@@ -2305,8 +2305,25 @@ _NODISCARD _OutputIt _Fmt_write(_OutputIt _Out, const _Arithmetic _Value) {
...
-    _STL_ASSERT(_Ec == errc{}, "to_chars failed");
+    char* _End = _Buffer;
+
+    if constexpr (is_floating_point_v<_Arithmetic>) {
+        if ((_STD isnan)(_Value)) {
+            if ((_STD signbit)(_Value)) {
+                *_End++ = '-';
+            }
+
+            memcpy(_End, "nan", 3);
```
Comment (StephanTLavavej) [inline stl/inc/format:2316]:
```suggestion
            _CSTD memcpy(_End, "nan", 3);
```

File: stl/inc/format [hunk 3]
```diff
@@ -2773,19 +2790,29 @@ _NODISCARD _OutputIt _Fmt_write(
...
-        _Result = _STD to_chars(_Buffer, _STD end(_Buffer), _Value, _Format);
+    const auto _Is_negative = (_STD signbit)(_Value);
+
+    if ((_STD isnan)(_Value)) {
+        _Result.ptr = _Buffer;
+        if (_Is_negative) {
+            ++_Result.ptr; // pretend to skip over a '-' that to_chars would put in _Buffer[0]
+        }
+
+        memcpy(_Result.ptr, "nan", 3);
```
Comment (StephanTLavavej) [inline stl/inc/format:2801]:
```suggestion
        _CSTD memcpy(_Result.ptr, "nan", 3);
```

File: tests/std/tests/P0645R10_text_formatting_formatting/test.cpp [hunk 4]
```diff
@@ -855,6 +879,7 @@ void test_float_specs() {
     assert(format(STR("{:06}"), Float{0}) == STR("000000"));
     assert(format(STR("{:06}"), Float{1.2}) == STR("0001.2"));
     assert(format(STR("{:06}"), nan) == STR("   nan"));
+    assert(format(STR("{:06}"), -nan) == STR("  -nan"));
     assert(format(STR("{:06}"), inf) == STR("   inf"));
```
Comment (StephanTLavavej) [inline tests/std/tests/P0645R10_text_formatting_formatting/test.cpp:883]: Pre-existing: Let's test `-inf` here to complete the symmetry.

File: tests/std/tests/P0645R10_text_formatting_formatting/test.cpp [hunk 5]
```diff
@@ -866,6 +891,7 @@ void test_float_specs() {
     assert(format(locale{"en-US"}, STR("{:.4Lf}"), value) == STR("1,234.5273"));
     assert(format(locale{"en-US"}, STR("{:#.4Lg}"), Float{0}) == STR("0.000"));
     assert(format(locale{"en-US"}, STR("{:L}"), nan) == STR("nan"));
+    assert(format(locale{"en-US"}, STR("{:L}"), -nan) == STR("-nan"));
     assert(format(locale{"en-US"}, STR("{:L}"), inf) == STR("inf"));
```
Comment (StephanTLavavej) [inline tests/std/tests/P0645R10_text_formatting_formatting/test.cpp:895]: Pre-existing: Ditto `-inf`.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
Thanks - this looks correct to me, we have sufficient buffer size, and all codepaths appear to be modified/tested. I pushed really trivial changes since I had to find *something*. :joy_cat:

**Comment by mordante** [comment]
Thanks for the quick fix @CaseyCarter

**Review by strega-nil-ms — APPROVED** [review]
Thanks @CaseyCarter 
excited akko!

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Comment by StephanTLavavej** [comment]
Thanks for fixing this bug! :lady_beetle: :abacus: :smile_cat:

**Pull request merged**